src/debug.sh

```diff
@@ -269,6 +278,233 @@ function event_trace()
+  local list="$7"
+  local disable="$8"
+  local raw_list=''
+  local save_following_log
+  local redirect_mode
+  local ret
+
+  if [[ -n "$list" || "$ftrace_syntax" == 1 ]]; then
+    ftrace_list "$target" "$flag"
+    return "$?"
+  fi
+
+  # Build basic trace command
+  cmd_ftrace=$(build_ftrace_command_string "$ftrace_syntax" "$disable")
+
+  # Capture data
+  if [[ -n "$base_log_path" ]]; then
+    touch "$base_log_path/ftrace"
+    printf '\n' > "$base_log_path/ftrace"
+    save_following_log="$base_log_path/ftrace"
+  fi
+
+  # Follow
+  if [[ "$follow" == 1 && -z "$disable" ]]; then
+    cmd_ftrace="$cmd_ftrace && cat $TRACE_PIPE"
+  fi
+
+  # Handling command
+  if [[ -n "$user_cmd" && -z "$disable" ]]; then
+    save_following_log="$base_log_path/ftrace"
+    screen_nick=$(get_today_info '+kw_%Y_%m_%d-%H_%M_%S')
+    screen_cmd="screen -L -Logfile ~/$screen_nick -dmS $screen_nick cat $TRACE_PIPE"
+    screen_end_cmd="screen -S $screen_nick -X quit > /dev/null"
+    disable_cmd=$(build_ftrace_command_string '' 1)
+    cmd_ftrace="$cmd_ftrace && $screen_cmd && $user_cmd && $disable_cmd && $screen_end_cmd"
+  fi
+
+  case "$target" in
+    2) # LOCAL
+      [[ -n "$save_following_log" ]] && redirect_mode='KW_REDIRECT_MODE'
+
+      cmd_manager "$flag" "$cmd_ftrace" "$redirect_mode" "$save_following_log"
+      ret="$?"
+
+      if [[ -n "$user_cmd" ]]; then
+        cmd_ftrace="sudo cp /root/$screen_nick $save_following_log && sudo chown $USER:$USER $save_following_log"
+        cmd_manager "$flag" "$cmd_ftrace"
+      fi
+      ;;
+    3 | 1) # REMOTE && VM
+      local remote="${remote_parameters['REMOTE_IP']}"
+      local port="${remote_parameters['REMOTE_PORT']}"
+      local user="${remote_parameters['REMOTE_USER']}"
+
+      if [[ "$target" == 1 ]]; then
+        say 'Target is a VM'
+        # TODO: We should check if the VM is up and running
+      fi
+
+      cmd_remotely "$cmd_ftrace" "$flag" "$remote" "$port" "$user" '' "$save_following_log"
+      ret="$?"
+
+      # If we used --cmd, we need to retrieve the log
+      if [[ -n "$user_cmd" ]]; then
+        cmd_ftrace="scp -P $port $user@$remote:~/$screen_nick $save_following_log"
+        cmd_manager "$flag" "$cmd_ftrace"
+      fi
+      ;;
+  esac
+
+  # 130 - Owner died happens during the interruption
+  if [[ "$ret" != 0 && "$ret" != 130 ]]; then
+    complain "Fail to enable ftrace: $ftrace_syntax - $ret"
+    complain 'Hint: try to use a wildcard in the filter'
+    return "$ret"
+  fi
+}
+
+# This function is responsible for building the ftrace string that represents
+# the command executed in the target machine.
+#
+# @ftrace_syntax
+#
+# Return:
+# A string with the final command is composed.
+function build_ftrace_command_string()
+{
+  local ftrace_syntax="$1"
+  local ftrace_disable="$2"
+  local char_repetition
+  local ftrace_filters
+  local ftrace_type
+  local cmd_disable_ftrace="echo 0 > $TRACING_ON"
+  local cmd_enable_ftrace="echo 1 > $TRACING_ON"
```

Comment:
Can't we use `printf` here? I don't know much about the kernel's debug feature, but if there are no limitations regarding this it would be better to stick to our codestyle.

Reply:
I decided to use `echo` in this case because this will be used in the ssh command in the target machine. This can simplify the debug and also reduce the side of the command a little bit.